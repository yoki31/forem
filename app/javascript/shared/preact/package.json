{
  "name": "preact-root-fragment",
  "version": "0.2.0",
  "main": "./preact-root-fragment.js",
  "module": "./preact-root-fragment.js",
  "type": "module"
}
