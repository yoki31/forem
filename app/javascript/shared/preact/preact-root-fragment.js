/**
 * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
 *
 * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
 * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
 * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
 */
export function createRootFragment(parent, replaceNode) {
  if (replaceNode) {
    replaceNode = Array.isArray(replaceNode) ? replaceNode : [replaceNode];
  } else {
    replaceNode = [parent];
    parent = parent.parentNode;
  }

  const s = replaceNode[replaceNode.length - 1].nextSibling;

  const rootFragment = {
    nodeType: 1,
    parentNode: parent,
    firstChild: replaceNode[0],
    childNodes: replaceNode,
    insertBefore: (c, r) => {
      parent.insertBefore(c, r || s);
      return c;
    },
    appendChild: (c) => {
      parent.insertBefore(c, s);
      return c;
    },
    removeChild: (c) => {
      parent.removeChild(c);
      return c;
    },
  };

  parent.__k = rootFragment;
  return rootFragment;
}
