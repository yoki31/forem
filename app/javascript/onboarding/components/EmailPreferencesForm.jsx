import { h, Component } from 'preact';
import PropTypes from 'prop-types';
import { getContentOfToken, updateOnboarding } from '../utilities';
import { Navigation } from './Navigation';

/* eslint-disable camelcase */
export class EmailPreferencesForm extends Component {
  constructor(props) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
    this.state = {
      content: '<p>Loading...</p>',
      askingToReconsiderEmail: false,
    };
  }

  componentDidMount() {
    fetch('/onboarding/newsletter')
      .then((response) => response.json())
      .then((json) => {
        this.setState({ content: json['content'] });
      });

    updateOnboarding('v2: email preferences form');
  }

  onSubmit() {
    const csrfToken = getContentOfToken('csrf-token');
    const newsletterEl = document.getElementById('email_newsletter');
    const newsletterChecked = newsletterEl ? newsletterEl.checked : false

    if (newsletterChecked) {
      fetch('/onboarding/notifications', {
        method: 'PATCH',
        headers: {
          'X-CSRF-Token': csrfToken,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ notifications: { email_newsletter: newsletterChecked } }),
        credentials: 'same-origin',
      }).then((response) => {
        if (response.ok) {
          localStorage.setItem('shouldRedirectToOnboarding', false);
          const { next } = this.props;
          next();
        }
      });
    } else if (!this.state.askingToReconsiderEmail) {
      this.setState({
        askingToReconsiderEmail: true,
      });
    }
  }

  finishWithoutEmail = () => {
    localStorage.setItem('shouldRedirectToOnboarding', false);
    const { next } = this.props;
    next();
  }

  finishWithEmail = () => {
    const csrfToken = getContentOfToken('csrf-token');
    fetch('/onboarding/notifications', {
      method: 'PATCH',
      headers: {
        'X-CSRF-Token': csrfToken,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ notifications: { email_newsletter: true, email_digest_periodic: true } }),
      credentials: 'same-origin',
    }).then((response) => {
      if (response.ok) {
        localStorage.setItem('shouldRedirectToOnboarding', false);
        const { next } = this.props;
        next();
      }
    });
  }

  renderEmailReconsideration() {
    if(!this.state.askingToReconsiderEmail) {
      return '';
    }
    return (
      <div>
      <div style='position:absolute;left:0;right:0;top:0;bottom:0;background:black;opacity:0.8;z-index:99;' />
        <div className='crayons-card onboarding-inner-popover'>
          <p style='padding: 3vh 0 1.5vh;color:var(--base-60);'>
            ðŸ‘‹ One last check
          </p>
          <h2 className="crayons-heading crayons-heading--bold">
            We Recommend Subscribing to Emails
          </h2>
          <p style='padding: 4vh 0 1.5vh;color:var(--base-60);max-width:660px;margin:auto;line-height:135%;'>
            Newsletters are a part of keeping up with the pulse of the overall DEV ecosystem.
            <span style='display:inline-block'>It's easy to unsubscribe later if it's not for you.</span>
          </p>
          <div className="align-center" style="padding: 5vh 0;">
            <button className="inline-block m-4 c-btn c-btn--ghost" style="opacity:0.8;" onClick={this.finishWithoutEmail}>No thank you</button>
            <button className="inline-block m-4 c-btn c-btn--primary" onClick={this.finishWithEmail}>Count me in</button>
          </div>
        </div>
      </div>);
  }
      

  render() {
    const { prev, slidesCount, currentSlideIndex } = this.props;
    return (
      <div
        data-testid="onboarding-email-preferences-form"
        className="onboarding-main crayons-modal crayons-modal--large"
      >
        <div
          className="crayons-modal__box"
          role="dialog"
          aria-labelledby="title"
          aria-describedby="subtitle"
        >
          <div
            className="onboarding-content email-preferences-wrapper"
            // eslint-disable-next-line react/no-danger
            dangerouslySetInnerHTML={{ __html: this.state.content }}
          />
          {this.renderEmailReconsideration()}
          <Navigation
            prev={prev}
            next={this.onSubmit}
            slidesCount={slidesCount}
            currentSlideIndex={currentSlideIndex}
          />
        </div>
      </div>
    );
  }
}

EmailPreferencesForm.propTypes = {
  prev: PropTypes.func.isRequired,
  next: PropTypes.string.isRequired,
  slidesCount: PropTypes.number.isRequired,
  currentSlideIndex: PropTypes.func.isRequired,
};

/* eslint-enable camelcase */
