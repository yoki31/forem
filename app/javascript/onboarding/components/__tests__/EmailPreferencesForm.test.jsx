import { h } from 'preact';
import { render, fireEvent, waitFor } from '@testing-library/preact';
import fetch from 'jest-fetch-mock';
import { axe } from 'jest-axe';

import { EmailPreferencesForm } from '../EmailPreferencesForm';

global.fetch = fetch;

describe('EmailPreferencesForm', () => {
  const renderEmailPreferencesForm = () =>
    render(
      <EmailPreferencesForm
        next={jest.fn()}
        prev={jest.fn()}
        currentSlideIndex={4}
        slidesCount={5}
        communityConfig={{
          communityName: 'Community Name',
          communityLogo: '/x.png',
          communityBackgroundColor: '#FFF000',
          communityDescription: 'Some community description',
        }}
        previousLocation={null}
      />,
    );

  const getUserData = () =>
    JSON.stringify({
      followed_tag_names: ['javascript'],
      profile_image_90: 'mock_url_link',
      name: 'firstname lastname',
      username: 'username',
    });

  const fakeResponse = JSON.stringify({
    content: `
    <h1>Almost there!</h1>
    <form>
      <fieldset>
        <ul>
          <li class="checkbox-item">
            <label for="email_newsletter"><input type="checkbox" id="email_newsletter" name="email_newsletter">I want to receive weekly newsletter emails.</label>
          </li>
        </ul>
      </fieldset>
    </form>
    `,
  });

  beforeEach(() => {
    fetch.resetMocks();
    fetch.mockResponseOnce(fakeResponse);
  });

  beforeAll(() => {
    document.head.innerHTML =
      '<meta name="csrf-token" content="some-csrf-token" />';
    document.body.setAttribute('data-user', getUserData());
  });

  it('should have no a11y violations', async () => {
    const { container } = render(renderEmailPreferencesForm());
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('should load the appropriate text', async () => {
    const { findByLabelText } = renderEmailPreferencesForm();
    await findByLabelText(/receive weekly newsletter/i);
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('should show the checkbox unchecked', async () => {
    const { findByLabelText } = renderEmailPreferencesForm();
    const checkbox = await findByLabelText(/receive weekly newsletter/i);
    expect(checkbox.checked).toBe(false);
  });

  it('should render a stepper', () => {
    const { queryByTestId } = renderEmailPreferencesForm();
    expect(queryByTestId('stepper')).not.toBeNull();
  });

  it('should render a back button', () => {
    const { queryByTestId } = renderEmailPreferencesForm();
    expect(queryByTestId('back-button')).not.toBeNull();
  });

  it('should render a button that says Finish', () => {
    const { queryByText } = renderEmailPreferencesForm();
    expect(queryByText('Finish')).not.toBeNull();
  });

  it('should show the reconsideration prompt when the checkbox is not checked', async () => {
    const { getByText } = renderEmailPreferencesForm();
    const finishButton = getByText('Finish');

    fireEvent.click(finishButton);

    await waitFor(() => {
      expect(getByText(/We Recommend Subscribing to Emails/i)).not.toBeNull();
    });
  });

  it('should handle "No thank you" button click in the reconsideration prompt', async () => {
    const { getByText, findByLabelText } = renderEmailPreferencesForm();
    const checkbox = await findByLabelText(/receive weekly newsletter/i);
    const finishButton = getByText('Finish');

    fireEvent.click(finishButton);

    await waitFor(() => {
      expect(getByText(/We Recommend Subscribing to Emails/i)).not.toBeNull();
    });

    const noThankYouButton = getByText('No thank you');
    fireEvent.click(noThankYouButton);

    // Verify that the `finishWithoutEmail` function is called
    expect(checkbox.checked).toBe(false);
  });

  it('should handle "Count me in" button click in the reconsideration prompt', async () => {
    const { getByText } = renderEmailPreferencesForm();
    const finishButton = getByText('Finish');

    fireEvent.click(finishButton);

    await waitFor(() => {
      expect(getByText(/We Recommend Subscribing to Emails/i)).not.toBeNull();
    });

    const countMeInButton = getByText('Count me in');
    fireEvent.click(countMeInButton);

    // Verify that the `finishWithEmail` function is called
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith('/onboarding/notifications', expect.any(Object));
    });
  });
});
