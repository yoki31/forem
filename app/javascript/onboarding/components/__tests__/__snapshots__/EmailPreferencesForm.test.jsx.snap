// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailPreferencesForm should load the appropriate text 1`] = `
"<div><div data-testid=\\"onboarding-email-preferences-form\\" class=\\"onboarding-main crayons-modal crayons-modal--large\\"><div class=\\"crayons-modal__box\\" role=\\"dialog\\" aria-labelledby=\\"title\\" aria-describedby=\\"subtitle\\"><div class=\\"onboarding-content email-preferences-wrapper\\">
    <h1>Almost there!</h1>
    <form>
      <fieldset>
        <ul>
          <li class=\\"checkbox-item\\">
            <label for=\\"email_newsletter\\"><input type=\\"checkbox\\" id=\\"email_newsletter\\" name=\\"email_newsletter\\">I want to receive weekly newsletter emails.</label>
          </li>
        </ul>
      </fieldset>
    </form>
    </div><nav class=\\"onboarding-navigation\\"><div class=\\"navigation-content\\"><div class=\\"back-button-container \\"><button data-testid=\\"back-button\\" class=\\"back-button\\" type=\\"button\\" aria-label=\\"Back to previous onboarding step\\"><svg width=\\"24\\" height=\\"24\\" fill=\\"none\\" class=\\"crayons-icon\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M7.828 11H20v2H7.828l5.364 5.364-1.414 1.414L4 12l7.778-7.778 1.414 1.414L7.828 11z\\"></path></svg></button></div><div data-testid=\\"stepper\\" class=\\"stepper\\"><span class=\\"dot active\\"></span><span class=\\"dot active\\"></span><span class=\\"dot active\\"></span><span class=\\"dot active\\"></span><span class=\\"dot active\\"></span></div><button class=\\"next-button\\" type=\\"button\\">Finish</button></div></nav></div></div></div>"
`;
